### PR TITLE
New time panel density graph

### PR DESCRIPTION
**Pull request by emilk**
Closes https://github.com/rerun-io/rerun/issues/619

The time panel used to be `O(N)` over the number of log messages. With this PR it is now `O(log(N))`.

In order to accomplish this I decided to replace the "ball scatterer" visualization of the time panel with a brand new density graph. The end result is something that is much faster, and also looks much better.

density-graph

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)

### CODE REVIEW

File: crates/re_int_histogram/tests/snapshots/memory_test__Int64Histogram.snap [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
 ---
 [
-    "5.7 B/entry, dense",
-    "35.5 B/entry, spacing: 1M",
-    "25.8 B/entry, spacing: 2M",
-    "25.3 B/entry, spacing: 3M",
-    "27.9 B/entry, spacing: 5M",
-    "35.5 B/entry, spacing: 8M",
-    "23.0 B/entry, spacing: 13M",
+    "9.6 B/entry, dense",
```
Comment (emilk) [inline crates/re_int_histogram/tests/snapshots/memory_test__Int64Histogram.snap:7]: I switched to a faster but more memory-intense variant of re_int_histogram

File: crates/re_int_histogram/src/tree.rs [hunk 2]
```diff
@@ -667,150 +700,203 @@ impl<'a> Iterator for TreeIterator<'a> {
...
-    debug_assert_eq!(set.min_key(), None);
-    debug_assert_eq!(set.max_key(), None);
-    let mut expected_ranges = vec![];
-    for i in 0..100 {
-        debug_assert_eq!(set.total_count(), i);
-        debug_assert_eq!(set.range_count(-10000..10000), i);
-        let key = i as i64;
-        set.increment(key, 1);
+#[cfg(test)]
+mod tests {
```
Comment (emilk) [inline crates/re_int_histogram/src/tree.rs:704]: image

File: crates/re_viewer/src/ui/time_panel/time_axis.rs [hunk 3]
```diff
@@ -113,18 +99,58 @@ fn gap_size_heuristic<T>(time_type: TimeType, times: &BTreeMap<TimeInt, T>) -> u
     gap_threshold
 }
 
-/// Collapse any gaps larger or equals to the given threshold.
-fn create_ranges<T>(times: &BTreeMap<TimeInt, T>, gap_threshold: u64) -> vec1::Vec1<TimeRange> {
+fn collect_gaps(times: &TimeHistogram, min_gap_size: u64, max_collapses: usize) -> Vec<u64> {
```
Comment (jleibs) [inline crates/re_viewer/src/ui/time_panel/time_axis.rs:102]:
```suggestion
fn collect_candidate_gaps(times: &TimeHistogram, min_gap_size: u64, max_collapses: usize) -> Vec<u64> {
```
Maybe rename to make this more clear these gaps won't necessarily be used for splits.

**Review by jleibs — APPROVED** [review]
Overall looks good. Couple small comment/cleanup possibilities.

File: crates/re_int_histogram/src/tree.rs [hunk 4]
```diff
@@ -13,21 +13,25 @@ use crate::{i64_key_from_u64_key, u64_key_from_i64_key, RangeI64, RangeU64};
...
 // Uses 20x nodes with 8-way (3 bit) branching factor down to a final 16-way (4 bit) dense leaf.
 // 20x 3-bit + 4-bit = 64 bit.
 // level 1, 4, 7, …, 58, 61
 // This uses about half the memory of 16-way branching, but is also half as fast.
 // 5.7 B/dense entry
 // 25-35 B/sparse entry
 
-/// How many bits we progress in each [`BranchNode`]
-const LEVEL_STEP: u64 = 3;
+// /// How many bits we progress in each [`BranchNode`]
```
[the remark this answers is not in the document]
Comment (emilk) [reply crates/re_int_histogram/src/tree.rs]: We decide based on whether we prioritize RAM or CPU. I turned them into `mod`s instead; looks a lot cleaner

File: crates/re_viewer/src/ui/time_panel/data_density_graph.rs [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+        // first bucket, partially filled:
+        if let Ok(i) = usize::try_from(min_bucket.floor() as i64) {
+            if let Some(bucket) = self.buckets.get_mut(i) {
+                *bucket += first_bucket_factor * count_per_bucket;
+            }
+        }
+
+        // full buckets:
+        for i in (min_bucket.ceil() as i64)..=(max_bucket.floor() as i64) {
+            if 0 <= i {
```
[the remark this answers is not in the document]
Comment (emilk) [reply crates/re_viewer/src/ui/time_panel/data_density_graph.rs:175]: We use `usize::try_from` there, but let's use it everywhere for conistency

**Pull request merged**